### PR TITLE
gnrc_netif: replace GNRC_NETIF_SINGLE with gnrc_netif_single pseudo-module

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

To being able to use this information during dependency resolution, make gnrc_netif_single a pseudo-module.

### Testing procedure

File size of the affected examples should not change.

### Issues/PRs references

https://github.com/RIOT-OS/RIOT/pull/13746#discussion_r461458518

### CODE REVIEW

File: sys/include/net/gnrc/netif/conf.h [hunk 1]
```diff
@@ -31,18 +31,6 @@
...
 #endif
 
-/**
- * @brief   Single interface optimizations
- *
- *          Define to 1 to allow GNRC optimizations when only one interface
- *          is available.
- *
- * @note    This MUST NOT be enabled if there's more than one interface.
- */
```
Comment (benpicco) [inline sys/include/net/gnrc/netif/conf.h:41]: I'm afraid we have no way of documenting pseudo-modules :(

**Comment by miri64** [comment]
Alternatively you could move this to a pseudo-module to `gnrc_netif_single`, but that won't help you with the issue I pointed out in  #13746.

File: sys/include/net/gnrc/netif/conf.h (hunk 1 above)
Comment (miri64) [reply sys/include/net/gnrc/netif/conf.h:41]: You could move it to the detailed description in `net/gnrc/netif.h`.

**Comment by miri64** [comment]
> […] but that won't help you with the issue I pointed out in #13746.

Ok, it could, since it makes it more obvious to the build system... ^^

File: sys/include/net/gnrc/netif.h [hunk 2]
```diff
@@ -330,7 +330,7 @@ unsigned gnrc_netif_numof(void);
  */
 static inline bool gnrc_netif_highlander(void)
 {
-    return IS_ACTIVE(GNRC_NETIF_SINGLE);
+    return IS_ACTIVE(MODULE_NETIF_SINGLE);
```
Comment (miri64) [inline sys/include/net/gnrc/netif.h:333]:
```suggestion
    return IS_USED(MODULE_NETIF_SINGLE);
```

**Comment by benpicco** [comment]
Ok, I re-wrote the PR to use `gnrc_netif_single` instead.

File: sys/include/net/gnrc/netif/conf.h (hunk 1 above)
Comment (benpicco) [reply sys/include/net/gnrc/netif/conf.h:41]: Ok.

File: tests/emcute/Makefile [hunk 3]
```diff
@@ -25,7 +25,7 @@ USEMODULE += sock_util
 
 CFLAGS += -DEMCUTE_TOPIC_MAXLEN="249"   # 256 - 7
 CFLAGS += -DSTDIO_UART_RX_BUFSIZE="512" # Adapt to SHELL_BUFSIZE in app
-CFLAGS += -DGNRC_NETIF_SINGLE           # Only one interface used and it makes
```
Comment (miri64) [inline tests/emcute/Makefile:28]: This should be moved to the other `USEMODULE`s now.

File: tests/gnrc_sock_dns/Makefile [hunk 4]
```diff
@@ -4,7 +4,7 @@ RIOTBASE ?= $(CURDIR)/../..
 
 export TAP ?= tap0
 
-CFLAGS += -DGNRC_NETIF_SINGLE           # Only one interface used and it makes
```
Comment (miri64) [inline tests/gnrc_sock_dns/Makefile:7]: Same here

File: tests/gnrc_tcp/Makefile [hunk 5]
```diff
@@ -13,7 +13,7 @@ TIMEOUT_US ?= 3000000
 TEST_ON_CI_BLACKLIST += all
 
 CFLAGS += -DSHELL_NO_ECHO
-CFLAGS += -DGNRC_NETIF_SINGLE           # Only one interface used and it makes
```
Comment (miri64) [inline tests/gnrc_tcp/Makefile:16]: And here

File: examples/emcute_mqttsn/Makefile [hunk 6]
```diff
@@ -29,7 +29,7 @@ ifneq (,$(EMCUTE_ID))
 endif
 
 # Optimize network stack to for use with a single network interface
-CFLAGS += -DGNRC_NETIF_SINGLE
+USEMODULE += gnrc_netif_single
```
Comment (miri64) [inline examples/emcute_mqttsn/Makefile:32]: ... and here.

File: examples/gnrc_minimal/Makefile [hunk 7]
```diff
@@ -32,9 +32,9 @@ CFLAGS += -DCONFIG_GNRC_NETIF_IPV6_ADDRS_NUMOF=2 \
           -DGNRC_NETIF_IPV6_GROUPS_NUMOF=2 -DCONFIG_GNRC_IPV6_NIB_NUMOF=1 \
           -DCONFIG_GNRC_IPV6_NIB_OFFL_NUMOF=1 # be able to configure at least one route
 
-# Enable single interface optimization. Set to 0 if more than one interface is
+# Enable single interface optimization. Remove this if more than one interface is
 # present
-CFLAGS += -DGNRC_NETIF_SINGLE=1
+USEMODULE += gnrc_netif_single
```
Comment (miri64) [inline examples/gnrc_minimal/Makefile:37]: and finally here.

**Comment by miri64** [comment]
I'm fine with the current state, however the makefiles should not desolve into chaos ;-)

File: examples/gnrc_minimal/Makefile (hunk 7 above)
Comment (benpicco) [reply examples/gnrc_minimal/Makefile]: Order has been restored.

File: examples/gnrc_minimal/Makefile [hunk 8]
```diff
@@ -22,6 +22,9 @@ USEMODULE += gnrc_ipv6
 USEMODULE += gnrc_icmpv6_echo
 # Use minimal standard PRNG
 USEMODULE += prng_minstd
+# Enable single interface optimization.
+# Remove this if more than one interface is present
+USEMODULE += gnrc_netif_single
```
Comment (miri64) [inline examples/gnrc_minimal/Makefile:27]: Move to the other GNRC modules?

Comment (benpicco) [reply examples/gnrc_minimal/Makefile]: Ok

**Comment by miri64** [comment]
Tested so far

- [ ] examples/emcute_mqttsn
- [ ] examples/gnrc_minimal
- [ ] tests/emcute
- [ ] tests/gnrc_sock_dns
- [ ] tests/gnrc_tcp

**Comment by miri64** [comment]
`tests/emcute` fails, but in the same way it fails in master. IIRC it worked for the release tests, so need to investigate that :eyes:

**Comment by miri64** [comment]
> `tests/emcute` fails, but in the same way it fails in master. IIRC it worked for the release tests, so need to investigate that :eyes:

Something in a274ea45fc5b483ea41ae0d1f1993898db30d32a (https://github.com/RIOT-OS/RIOT/pull/12428) broke that test according to my bisect :-/

**Comment by miri64** [comment]
> Something in [a274ea4](https://github.com/RIOT-OS/RIOT/commit/a274ea45fc5b483ea41ae0d1f1993898db30d32a) (#12428) broke that test according to my bisect :-/

as the failure happens on the large payload test, I suspect something going wrong there.

**Comment by miri64** [comment]
> as the failure happens on the large ~payload~ _topic name_ test, I suspect something going wrong there.

No, if I remove the previous two test cases the test case succeeds, but then the last one fails, so its always on the third test case :-/

**Comment by miri64** [comment]
See https://github.com/RIOT-OS/RIOT/issues/14636

**Pull request merged**